dlt/destinations/sql_client.py

```diff
@@ -316,18 +341,54 @@ def df(self, chunk_size: int = None, **kwargs: Any) -> Optional[DataFrame]:
         May use native pandas/arrow reader if available. Depending on
         the native implementation chunk size may vary.
         """
-        from dlt.common.libs.pandas_sql import _wrap_result
+        try:
+            return next(self.iter_df(chunk_size=chunk_size))
+        except StopIteration:
+            return None
 
-        columns = self._get_columns()
-        if chunk_size is None:
-            return _wrap_result(self.native_cursor.fetchall(), columns, **kwargs)
-        else:
-            df = _wrap_result(self.native_cursor.fetchmany(chunk_size), columns, **kwargs)
-            # if no rows return None
-            if df.shape[0] == 0:
-                return None
-            else:
-                return df
+    def arrow(self, chunk_size: int = None, **kwargs: Any) -> Optional[ArrowTable]:
+        """Fetches results as data frame in full or in specified chunks.
+
+        May use native pandas/arrow reader if available. Depending on
+        the native implementation chunk size may vary.
+        """
+        try:
+            return next(self.iter_arrow(chunk_size=chunk_size))
+        except StopIteration:
+            return None
+
+    def iter_fetchmany(self, chunk_size: int) -> Generator[List[Tuple[Any, ...]], Any, Any]:
+        while True:
+            if not (result := self.fetchmany(chunk_size)):
+                return
+            yield result
+
+    def iter_df(self, chunk_size: int) -> Generator[DataFrame, None, None]:
+        """Default implementation converts arrow to df"""
+        from dlt.common.libs.pandas import pandas as pd
+
+        for table in self.iter_arrow(chunk_size=chunk_size):
+            # NOTE: we go via arrow table, types are created for arrow is columns are known
+            # https://github.com/apache/arrow/issues/38644 for reference on types_mapper
+            yield table.to_pandas(types_mapper=pd.ArrowDtype)
+
+    def iter_arrow(self, chunk_size: int) -> Generator[ArrowTable, None, None]:
+        """Default implementation converts query result to arrow table"""
+        from dlt.common.libs.pyarrow import row_tuples_to_arrow
+
+        # if loading to a specific pipeline, it would be nice to have the correct caps here
+        caps = DestinationCapabilitiesContext.generic_capabilities()
```

Comment:
all `sql_client` and `clients` have caps. this must be present at this point if we create dataset out of the `pipeline`. so assume you have caps in `init` of the cursor. but if it is None, you use generic caps.

also note that caps are often dynamic ie. filesystem + delta generates caps

Reply:
As far as I understand these capabilities are needed to create an arrow table that can be properly loaded into the destination that it will go. Since this is a read interface, it does not make sense to use the capabilities of the source database. If we know where the data is going (from one pipeline into the next one) the capabilities of the destination pipeline should be known here and used for creating the arrow tables, this is what I wrote in the comment above. Or is there some benefit of using the ones of the datasource?

Reply:
I added this plus a fallback to the defaults